File: output/objc/AddressLenses.m
#import "AddressLenses.h"

@implementation AddressLenses

+ (Lens<Address*, NSNumber*>*)postalCode
{
    return [Lens lensWithGet:^NSNumber*(Address* whole){
        return whole.postalCode;
    } set:^Address*(Address* whole, NSNumber* part) {
        return [whole addressBySettingPostalCode:part];
    }];
}
+ (Lens<Address*, NSString*>*)streetAddress
{
    return [Lens lensWithGet:^NSString*(Address* whole){
        return whole.streetAddress;
    } set:^Address*(Address* whole, NSString* part) {
        return [whole addressBySettingStreetAddress:part];
    }];
}
+ (Lens<Address*, NSNumber*>*)number
{
    return [Lens lensWithGet:^NSNumber*(Address* whole){
        return @(whole.number);
    } set:^Address*(Address* whole, NSNumber* part) {
        return [whole addressBySettingNumber:part.integerValue];
    }];
}
+ (Lens<Address*, NSNumber*>*)valid
{
    return [Lens lensWithGet:^NSNumber*(Address* whole){
        return @(whole.valid);
    } set:^Address*(Address* whole, NSNumber* part) {
        return [whole addressBySettingValid:part.boolValue];
    }];
}

@end

@implementation BoundLensToAddress

- (BoundLens*)postalCode 
{
    return [BoundLens lensWithParent:self sublens:[AddressLenses postalCode]];
}
- (BoundLens*)streetAddress 
{
    return [BoundLens lensWithParent:self sublens:[AddressLenses streetAddress]];
}
- (BoundLens*)number 
{
    return [BoundLens lensWithParent:self sublens:[AddressLenses number]];
}
- (BoundLens*)valid 
{
    return [BoundLens lensWithParent:self sublens:[AddressLenses valid]];
}

@end

@implementation Address (Lens)

- (BoundLensToAddress*)lens
{
    return [BoundLensToAddress lensWithInstance:self lens:[Lens identityLens]];
}

@end

File: output/swift/AddressLenses.swift
import Foundation

extension Address {
    struct Lenses {
        static let postalCode: Lens<Address, Int> = Lens(
            get: { $0.postalCode },
            set: { (address, postalCode) in
                var builder = AddressBuilder(existingAddress: address)
                return builder.withPostalCode(postalCode).build()
            }
        )
        static let streetAddress: Lens<Address, String> = Lens(
            get: { $0.streetAddress },
            set: { (address, streetAddress) in
                var builder = AddressBuilder(existingAddress: address)
                return builder.withStreetAddress(streetAddress).build()
            }
        )
        static let number: Lens<Address, Int> = Lens(
            get: { $0.number },
            set: { (address, number) in
                var builder = AddressBuilder(existingAddress: address)
                return builder.withNumber(number).build()
            }
        )
    }
}

struct BoundLensToAddress<Whole>: BoundLensType {
    typealias Part = Address
    let storage: BoundLensStorage<Whole, Part>
    
    var postalCode: BoundLens<Whole, Int> {
        return BoundLens<Whole, Int>(parent: self, sublens: Address.Lenses.postalCode)
    }

    var streetAddress: BoundLens<Whole, String> {
        return BoundLens<Whole, String>(parent: self, sublens: Address.Lenses.streetAddress)
    }

    var number: BoundLens<Whole, Int> {
        return BoundLens<Whole, Int>(parent: self, sublens: Address.Lenses.number)
    }
}

extension Address {
    var lens: BoundLensToAddress<Address> {
        return BoundLensToAddress<Address>(instance: self, lens: createIdentityLens())
    }
}
